### PR TITLE
[DependencyInjection] Add section about Service Closures

### PR DESCRIPTION
**Pull request by HypeMC**
Documents service closures added in https://github.com/symfony/symfony/pull/21770 and https://github.com/symfony/symfony/pull/41176, and the `service_closure()` PHP-DSL function added in https://github.com/symfony/symfony/pull/42625 .

### CODE REVIEW

File: service_container/service_closures.rst [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+            // $services->set(MyService::class)
+            //     ->args([service_closure('mailer')->ignoreOnInvalid()]);
+        };
+
+.. seealso::
+
+    Another way to inject services lazily is via a :doc:`service locators </service_container/service_subscribers_locators>`.
+
+Using Service Closures in Compiler Passes
+-----------------------------------------
```
Comment (OskarStark) [inline service_container/service_closures.rst]:
```suggestion
Using Service Closures in Compiler Pass
---------------------------------------
```

File: service_container/service_subscribers_locators.rst [hunk 2]
```diff
@@ -12,6 +12,10 @@ instantiation of the services to be lazy. However, that's not possible using
 the explicit dependency injection since services are not all meant to
 be ``lazy`` (see :doc:`/service_container/lazy_services`).
 
+.. seealso::
+
+    Another way to inject services lazily is via a :doc:`service closure </service_container/service_closures>`.
```
Comment (OskarStark) [inline service_container/service_subscribers_locators.rst]:
```suggestion
    Another way to inject services lazily is via a
    :doc:`service closure </service_container/service_closures>`.
```

File: service_container/service_closures.rst [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+                ->args([service_closure('mailer')]);
+
+            // In case the dependency is optional
+            // $services->set(MyService::class)
+            //     ->args([service_closure('mailer')->ignoreOnInvalid()]);
+        };
+
+.. seealso::
+
+    Another way to inject services lazily is via a :doc:`service locators </service_container/service_subscribers_locators>`.
```
Comment (OskarStark) [inline service_container/service_closures.rst]:
```suggestion
    Another way to inject services lazily is via a
    :doc:`service locators </service_container/service_subscribers_locators>`.
```

File: service_container/service_closures.rst [hunk 4]
```diff
@@ -0,0 +1,112 @@
+.. index::
+    single: DependencyInjection; Service Closures
+
+Service Closures
+================
+
```
Comment (OskarStark) [inline service_container/service_closures.rst:6]:
```suggestion

.. versionadded:: 5.4

    Service closures were introduced in Symfony 5.4.

```

File: service_container/service_closures.rst [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    .. code-block:: xml
+
+        <!-- config/services.xml -->
+        <?xml version="1.0" encoding="UTF-8" ?>
+        <container xmlns="http://symfony.com/schema/dic/services"
+            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+            xsi:schemaLocation="http://symfony.com/schema/dic/services https://symfony.com/schema/dic/services/services-1.0.xsd">
+
+            <services>
+
```
Comment (OskarStark) [inline service_container/service_closures.rst]: I would remove this blank line

File: service_container/service_closures.rst [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+                <service id="App\Service\MyService">
+                    <argument type="service_closure" id="mailer"/>
+                </service>
+
```
Comment (OskarStark) [inline service_container/service_closures.rst]: I would remove this blank line

File: service_container/service_closures.rst (hunk 4 above)
Comment (HypeMC) [reply service_container/service_closures.rst:6]: @OskarStark Service closures were introduced in version 3.3, only the `service_closure()` DSL function was added in 5.4

File: service_container/service_closures.rst [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        public function doSomething(): void
+        {
+            // ...
+
+            $this->getMailer()->send($email);
+        }
+
+        private function getMailer(): MailerInterface
+        {
+            return ($this->mailer)();
```
Comment (OskarStark) [inline service_container/service_closures.rst:45]: Do we need the getMailer() method or can we interact with the property directly?

File: service_container/service_closures.rst (hunk 4 above)
Comment (OskarStark) [reply service_container/service_closures.rst:6]: Ok, please adjust the sentence accordingly. Thanks

File: service_container/service_closures.rst (hunk 1 above)
Comment (OskarStark) [reply service_container/service_closures.rst]: Maybe even use Service Closure (singular) in the headline?

Comment (HypeMC) [reply service_container/service_closures.rst]: What about "Using a Service Closures in a Compiler Pass"?

File: service_container/service_subscribers_locators.rst (hunk 2 above)
Comment (HypeMC) [reply service_container/service_subscribers_locators.rst]: Done

File: service_container/service_closures.rst (hunk 3 above)
Comment (HypeMC) [reply service_container/service_closures.rst]: Done

File: service_container/service_closures.rst (hunk 4 above)
Comment (HypeMC) [reply service_container/service_closures.rst:6]: Added

File: service_container/service_closures.rst (hunk 5 above)
Comment (HypeMC) [reply service_container/service_closures.rst]: Done

File: service_container/service_closures.rst (hunk 6 above)
Comment (HypeMC) [reply service_container/service_closures.rst]: Done

File: service_container/service_closures.rst (hunk 7 above)
Comment (HypeMC) [reply service_container/service_closures.rst:45]: No, with the improved phpdoc, this becomes useless, removed.

Comment (HeahDude) [reply service_container/service_closures.rst:45]: I would have kept it, the phpdocs helps understanding the code when crossing the private closure, but this method ensures the expected type is returned at runtime.

Comment (HypeMC) [reply service_container/service_closures.rst:45]: I'm OK either way, so when you come to an agreement, please let me know and I'll update the PRs.

Comment (HypeMC) [reply service_container/service_closures.rst:45]: Hi guys, I was wondering if you've reached any consensus regarding this one?

Comment (OskarStark) [reply service_container/service_closures.rst:45]: Lets go with @HeahDude's comment then, easier way to understand is always a plus for me

Comment (HypeMC) [reply service_container/service_closures.rst:45]: Done

**Review by OskarStark — APPROVED** [review]

File: service_container/lazy_services.rst [hunk 8]
```diff
@@ -6,7 +6,8 @@ Lazy Services
 
 .. seealso::
 
-    Another way to inject services lazily is via a :doc:`service subscriber </service_container/service_subscribers_locators>`.
+    Other ways to inject services lazily is via a :doc:`service closure </service_container/service_closures>` or
```
Comment (jdreesen) [inline service_container/lazy_services.rst]:
```suggestion
    Other ways to inject services lazily are via a :doc:`service closure </service_container/service_closures>` or
```

File: service_container/service_closures.rst [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+
+            // In case the dependency is optional
+            // $services->set(MyService::class)
+            //     ->args([service_closure('mailer')->ignoreOnInvalid()]);
+        };
+
+.. seealso::
+
+    Another way to inject services lazily is via a
+    :doc:`service locators </service_container/service_subscribers_locators>`.
```
Comment (jdreesen) [inline service_container/service_closures.rst]:
```suggestion
    :doc:`service locator </service_container/service_subscribers_locators>`.
```

File: service_container/service_closures.rst [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+
+Using a Service Closures in a Compiler Pass
```
Comment (jdreesen) [inline service_container/service_closures.rst]:
```suggestion
Using a Service Closure in a Compiler Pass
```

**Review by OskarStark — APPROVED** [review]

**Comment by OskarStark** [comment]
Thank you @HypeMC.

File: service_container/service_closures.rst [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+
+        use App\Service\MyService;
+
+        return function (ContainerConfigurator $containerConfigurator) {
+            $services = $containerConfigurator->services();
+
+            $services->set(MyService::class)
+                ->args([service_closure('mailer')]);
+
+            // In case the dependency is optional
```
Comment (HeahDude) [inline service_container/service_closures.rst:89]: This comment should be either removed or added in YAML and XML as well for the consistency of the config block.

**Review by HeahDude — COMMENTED** [review]
Good job with this PR, I've left a small comment, sorry for the delay.

File: service_container/service_closures.rst (hunk 11 above)
Comment (OskarStark) [reply service_container/service_closures.rst:89]: See https://github.com/symfony/symfony-docs/pull/18050

**Pull request merged**